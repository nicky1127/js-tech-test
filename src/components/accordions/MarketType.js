import React, { useState } from "react";
import PropTypes from "prop-types";
import clsx from "clsx";
import { connect } from "react-redux";
import { makeStyles } from "@material-ui/core/styles";
import {
  Accordion,
  AccordionSummary,
  AccordionDetails,
  Typography,
  Paper,
  Table,
  TableContainer,
  TableRow,
  TableCell,
  TableBody
} from "@material-ui/core";
import { ExpandMore } from "@material-ui/icons";
import { getOutcomeListByIds } from "redux/_actions";
import labels from "constants/labels";

const constants = labels.MarketType;

const useStyles = makeStyles(theme => ({
  root: {
    margin: "10px"
  },
  header: {
    color: "#fff",
    backgroundColor: "rgb(10,40,108)",
    borderTopLeftRadius: "5px",
    borderTopRightRadius: "5px"
  },
  icon: { color: "#fff" },
  details: {
    padding: 0,
    flexDirection: "column"
  },
  table: {
    "& th": {
      border: "1px solid rgb(179,191,211)"
    },
    "& td": {
      border: "1px solid rgb(179,191,211)"
    },
    
  },
  suspendedCell: {
    backgroundColor: "rgb(205,205,206)",
    color: "rgb(120,120,120)",
    fontWeight: "600"
  },
  normalCell: {
    color: "rgb(217,34,49)",
    "&:hover": {
      backgroundColor: "rgb(239,244,252)",
      cursor: "pointer"
    }
  }
}));

export const MarketType = props => {
  const classes = useStyles();
  const { event, market, isOddsDecimal, getOutcomeListByIds } = props;
  const [expanded, setExpanded] = useState(false);

  const rows =
    market.outcomeList &&
    Object.values(market.outcomeList).map(outcome => ({
      status: outcome.status,
      name: outcome.name,
      price: outcome.price
    }));

  const onClickBtn = () => {
    setExpanded(!expanded);
    if (!Object.values(market.outcomeList).length) {
      getOutcomeListByIds(market.outcomes);
    }
  };

  return (
    <div className={clsx("marketType", classes.root)}>
      <Accordion expanded={expanded}>
        <AccordionSummary
          className={classes.header}
          expandIcon={<ExpandMore fontSize="large" className={classes.icon} />}
          onClick={onClickBtn}
          aria-controls="panel1a-content"
          id="panel1a-header"
        >
          <Typography className={classes.heading}>{market.name}</Typography>
        </AccordionSummary>
        <AccordionDetails classes={{ root: classes.details }}>
          <TableContainer component={Paper}>
            <Table className={classes.table} aria-label="simple table">
              <TableBody>
                {rows &&
                  rows.map(row => {
                    const price = isOddsDecimal
                      ? row.price.decimal
                      : `${row.price.num}/${row.price.den}`;
                    return (
                      <TableRow key={row.name}>
                        <TableCell component="th" scope="row">
                          {row.name}
                        </TableCell>
                        {row.status.suspended ? (
                          <TableCell
                            className={classes.suspendedCell}
                            align="center"
                          >
                            Susp
                          </TableCell>
                        ) : (
                          <TableCell
                            className={classes.normalCell}
                            align="center"
                          >
                            {price}
                          </TableCell>
                        )}
                      </TableRow>
                    );
                  })}
              </TableBody>
            </Table>
          </TableContainer>
        </AccordionDetails>
      </Accordion>
    </div>
  );
};

MarketType.defaultProps = { market: {} };

MarketType.propTypes = {};

const mapStateToProps = state => {
  const { isOddsDecimal } = state;
  return { isOddsDecimal };
};

const ConnectedMarketType = connect(mapStateToProps, { getOutcomeListByIds })(
  MarketType
);

export default ConnectedMarketType;
