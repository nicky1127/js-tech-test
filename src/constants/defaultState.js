const state = {
  liveEvents: [],
  event: {},
  markets: {},
  isOddsDecimal: false,
  loading: false,
  error: {},
  selections: []
};

export default state;
