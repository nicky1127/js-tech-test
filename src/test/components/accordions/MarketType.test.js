import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import MarketType from "components/accordions/MarketType";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import defaultState from "constants/defaultState";

beforeEach(() => {
  jest.clearAllMocks();
  cleanup();
});

const props = {};

const state = {
  ...defaultState
};
const mockStore = configureMockStore();
const store = mockStore(state);

describe("MarketType", () => {
  it("Snapshot matchs", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MarketType {...props} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("MarketType should render one [data-testid=marketType]", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <MarketType {...props} />
      </Provider>
    );
    const container = getByTestId("marketType");
    expect(container).toBeInTheDocument();
  });
});
