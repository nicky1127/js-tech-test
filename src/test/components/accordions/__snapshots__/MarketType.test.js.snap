// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketType Snapshot matchs 1`] = `
<div
  className="marketType makeStyles-root-1"
  data-testid="marketType"
>
  <div
    className="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-controls="panel1a-content"
      aria-disabled={false}
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-header-2"
      id="panel1a-header"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        />
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root makeStyles-icon-3 MuiSvgIcon-fontSizeLarge"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            id="panel1a-content"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root makeStyles-details-4"
            >
              <div
                className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
              >
                <table
                  aria-label="simple table"
                  className="MuiTable-root makeStyles-table-5"
                  role={null}
                >
                  <tbody
                    className="MuiTableBody-root"
                    role={null}
                  />
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
